### PR TITLE
feat: connection over rest

### LINKED ISSUE #4959: Demonstrate that PyFluent can run transparently over REST or gRPC for solver settings
### Step 1 (exploration)
 
my-rest-client ---> fluent REST API for solver settings
 
my-rest-client: **BASIC PYTHON**
 
### Step 2
 
my-simple-launcher ---> my-session-class ---> flobject ---> my-rest-proxy ---> my-rest-client
 
my-simple-launcher: **TRIVIAL**
my-session-class: **SMALL**
flobject: **UNCHANGED**
my-rest-proxy: **SUBS FOR gRPC/IN-PROCESS**

### PR DESCRIPTION
**Pull request by mayankansys**
## Context
PyFluent currently supports a gRPC transport. This PR adds a parallel REST(HTTP/HTTPS) transport layer to exercise the same settings API over HTTP/HTTPS against the REST web server, keeping the settings tree transport-agnostic.

This PR covers **Step 1** and **Step 2** of Issue #4959:

- **Step 1** — Introduce a working REST client that can read/write settings and invoke command/query endpoints.
- **Step 2** — Provide a connection/session so the existing settings tree can run over REST without changing the settings tree implementation.

## Change Summary
- Added a REST client that builds requests, hashes auth tokens into the Authorization header, parses errors, and supports retry on transient failures for safe methods.
- Added a REST connection method.
- Added integration tests (real-server) and mocked/unit tests to validate request construction and error handling in CI.

## Rationale
Keep REST support isolated, simple, and low risk:
- no changes to the settings tree core logic,
- REST and gRPC can coexist,
- CI can catch regressions without requiring a live server.

## Impact
**What changes for users**
- New public APIs to connect via REST (HTTP/HTTPS).
- Settings reads/writes and command execution can be performed over HTTP/HTTPS.

**What does NOT change**
- The settings tree core remains unchanged.
- Existing gRPC launch/session APIs remain unchanged.

## Limitations (out of scope)
- Some server behaviors (shutdown timing, confirmation prompts) are server-driven and may vary by version/config.

### CODE REVIEW

File: doc/changelog.d/5015.added.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Connection over rest
```
[the remark this answers is not in the document]
Comment (seanpearsonuk) [reply doc/changelog.d/5015.added.md:1]: 👍

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+
+* :func:`connect_to_webserver` – connects to an **already-running** SimBA
+  server.  Requires ``ip``, ``port``, and ``auth_token`` to be supplied
+  explicitly.  Performs a reachability probe before returning the session.
+
+Environment variables
+---------------------
+``FLUENT_WEBSERVER_TOKEN``
+    Bearer token (password) that the embedded SimBA server expects.
+    **Required** — set this variable before calling :func:`launch_webserver`.
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/rest/rest_launcher.py]: Can the client software handle this internally? Is it necessary to involve the end user in this?

File: src/ansys/fluent/core/rest/README.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+---
+
+## Environment Variables
+
+| Variable | Required | Description |
+|----------|----------|-------------|
+| `FLUENT_WEBSERVER_TOKEN` | **Yes** | Shared secret between client and Fluent web server |
+| `FLUENT_REST_PORT` | No | Port for integration tests against a live server |
+| `FLUENT_REST_HOST` | No | Host for integration tests (default: `127.0.0.1`) |
+| `PYFLUENT_FLUENT_ROOT` | No | Override path to Fluent installation |
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/rest/README.md]: Not sure when each of these is applicable. E.g., if the user is launching locally then the host is clear. If connecting then host and port are both local to the connection, rather than persistent, global environment variables. For launch, the client software could create an ad hoc token. For connection, the user needs to pass the known token but that should be local to that connection. `PYFLUENT_FLUENT_ROOT` is a broader developer variable.

File: src/ansys/fluent/core/rest/__init__.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  lightweight solver session that wires ``FluentRestClient`` into
+  ``flobject.get_root`` so the full settings tree works over HTTP.
+
+* :func:`~ansys.fluent.core.rest.rest_launcher.launch_webserver` – **primary
+  entry point**.  Spawns a local Fluent process with ``-ws -ws-port={port}``,
+  reads the mandatory ``FLUENT_WEBSERVER_TOKEN`` env var, and returns a
+  connected session.
+
+* :func:`~ansys.fluent.core.rest.rest_launcher.connect_to_webserver` –
+  connects to an already-running SimBA server using explicit ``ip``, ``port``,
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/rest/__init__.py]: Maybe SimBA can be seen as an implementation detail. Mentioning it could just be confusing for our user base.

File: src/ansys/fluent/core/rest/client.py [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Pure-Python REST client for the Fluent solver settings (DataModel API).
+
+Fluent embeds an HTTP server (SimBA - Simulation Bridge Application) that
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/rest/client.py]: I'd say that this is Fluent documentation and should be omitted.

File: src/ansys/fluent/core/rest/__init__.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+Example::
+
+    from ansys.fluent.core.rest import launch_webserver
+
+    session = launch_webserver()
+    print(session.get_var("setup/models/energy/enabled"))
+    session.set_var("setup/models/energy/enabled", False)
+    session.execute_command("file/read-case", file_name="elbow.cas.h5")
+    session.exit()
+"""
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply src/ansys/fluent/core/rest/__init__.py:38]: @mayankansys, here you just need to update the docstrings as you are using the client in the test and get rid of this extra layer like we have discussed. It is already there in code, just not implemented here.

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    from ansys.fluent.core.rest import launch_webserver, connect_to_webserver
+    session = launch_webserver()
+    session.get_var("setup/models/energy/enabled")
+    session.exit()
+
+Connect to an already-running web server with known IP, port, and auth token::
+
+    session = connect_to_webserver("127.0.0.1", <port>, auth_token=<auth_token>)
+    session.set_var("setup/models/energy/enabled", False)
+"""
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply src/ansys/fluent/core/rest/rest_launcher.py]: @mayankansys, same like the one before.

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+import ssl
+import subprocess
+import time
+import urllib.error
+import urllib.request
+
+from ansys.fluent.core.launcher.process_launch_string import get_fluent_exe_path
+from ansys.fluent.core.rest.client import FluentRestClient
+from ansys.fluent.core.rest.tls import _TlsCertificateManager
+
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply src/ansys/fluent/core/rest/rest_launcher.py]: @mayankansys, this is because of a missing import, but anyways this will no longer be there right?

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 9]
```diff
@@ -0,0 +1,721 @@
...
+# ---------------------------------------------------------------------------
+# Public API — launchers
+# ---------------------------------------------------------------------------
+
+
+def launch_webserver(
+    *,
+    product_version: str | None = None,
+    fluent_path: str | None = None,
+    dimension: str = "3ddp",
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply src/ansys/fluent/core/rest/rest_launcher.py]: @mayankansys, you could add a test for this (advisable) or ignore it as other tests will use it as well internally.

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 10]
```diff
@@ -0,0 +1,763 @@
...
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply src/ansys/fluent/core/rest/rest_launcher.py]: @mayankansys, please add a test for this.

File: tests/conftest.py [hunk 11]
```diff
@@ -521,3 +521,71 @@ def datamodel_api_version_all(request, monkeypatch: pytest.MonkeyPatch) -> None:
...
+_REST_TOKEN = os.environ.get("FLUENT_WEBSERVER_TOKEN", "")
+_REST_PORT_STR = os.environ.get("FLUENT_REST_PORT", "")
+_REST_HOST = os.environ.get("FLUENT_REST_HOST", "127.0.0.1")
+_REST_COMPONENT = os.environ.get("FLUENT_REST_COMPONENT", "fluent_1")
+_REST_SCHEME = os.environ.get("FLUENT_REST_SCHEME", "http")
+
+
+def _rest_env_vars_present() -> bool:
+    """Return ``True`` when mandatory REST env vars are set."""
+    return bool(_REST_TOKEN and _REST_PORT_STR)
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/conftest.py]: @mayankansys, all the previous 3 comments mention the same thing. Please have a look into this as this will be resolved if we change the location of setting the env var.

File: tests/conftest.py [hunk 12]
```diff
@@ -521,3 +521,71 @@ def datamodel_api_version_all(request, monkeypatch: pytest.MonkeyPatch) -> None:
...
+
+    url = f"{_REST_SCHEME}://{_REST_HOST}:{port}/api/connection/run_mode"
+    req = urllib.request.Request(url, method="GET")
+    req.add_header(
+        "Authorization",
+        f"Bearer {hashlib.sha256(_REST_TOKEN.encode()).hexdigest()}",
+    )
+    try:
+        with urllib.request.urlopen(req, timeout=3):  # nosec B310
+            return True
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/conftest.py]: @mayankansys, this we need to update for https support as well, like you have done in launcher already.

File: tests/conftest.py [hunk 13]
```diff
@@ -521,3 +521,71 @@ def datamodel_api_version_all(request, monkeypatch: pytest.MonkeyPatch) -> None:
...
+    """
+    from ansys.fluent.core.rest.client import FluentRestClient
+
+    if not _rest_env_vars_present():
+        pytest.skip(
+            "REST env vars not set — set FLUENT_WEBSERVER_TOKEN and "
+            "FLUENT_REST_PORT to run real-server tests."
+        )
+    if not _rest_server_reachable():
+        pytest.skip(f"REST server at {_REST_HOST}:{_REST_PORT_STR} not reachable.")
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/conftest.py]: @mayankansys, this is self-explanatory. Can be ignored in this particular case. However why are we not taking it as integer initially only?

File: tests/test_rest.py [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+        toggled = not original
+        real_client.set_var(path, toggled)
+        readback = real_client.get_var(path)
+        assert (
+            readback == toggled
+        ), f"set_var did not take effect: expected {toggled}, got {readback}"
+
+        # Restore
+        real_client.set_var(path, original)
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, can be ignored in test, but remember this pattern as it is useful during coding.

File: tests/test_rest.py [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+        restored = real_client.get_var(path)
+        assert restored == original
+
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: same comment as above.

File: tests/test_rest.py [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+
+class TestRealGetObjectNames:
+    """GET /api/fluent_1/{path} — returns dict with names as keys."""
+
+    def test_velocity_inlet_returns_string_list(self, real_client):
+        """Verify that a named-object container returns a list of strings."""
+        names = real_client.get_object_names("setup/boundary-conditions/velocity-inlet")
+        assert isinstance(names, list)
+        assert len(names) > 0
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+        assert all(isinstance(n, str) for n in names)
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+
+    def test_pressure_outlet_returns_list(self, real_client):
+        """Verify that another named-object container also returns a list."""
+        names = real_client.get_object_names(
+            "setup/boundary-conditions/pressure-outlet"
+        )
+        assert isinstance(names, list)
+        assert len(names) > 0
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+# ---------------------------------------------------------------------------
+
+
+class TestRealGetListSize:
+    """GET /api/fluent_1/{path} — count object keys."""
+
+    def test_velocity_inlet_size_positive(self, real_client):
+        """Verify that a named-object container has a positive size."""
+        size = real_client.get_list_size("setup/boundary-conditions/velocity-inlet")
+        assert isinstance(size, int)
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: tests/test_rest.py [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+        assert size > 0
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply tests/test_rest.py]: @mayankansys, this can be ignored as we are aware of the fixture we are loading.

File: pyproject.toml [hunk 23]
```diff
@@ -23,6 +23,7 @@ dependencies = [
     "ansys-tools-common>=0.4.0",
     "ansys-tools-filetransfer>=0.2,<1.0",
     "ansys-units~=0.11.0",
+    "cryptography>=43.0.0",
```
[the remark this answers is not in the document]
Comment (prmukherj) [reply pyproject.toml]: @mayankansys, this is important. Please have a look into it.

File: src/ansys/fluent/core/rest/client.py [hunk 24]
```diff
@@ -0,0 +1,763 @@
...
+
+    def execute_query(self, path: str, query: str, **kwds) -> Any:
+        """Execute *query* at *path* (POST {path}/{query})."""
+        return self._execute(path, query, **kwds)
+
+    # ------------------------------------------------------------------
+    # Session lifecycle
+    # ------------------------------------------------------------------
+
+    def exit(self) -> None:
```
Comment (prmukherj) [inline src/ansys/fluent/core/rest/client.py:358]: @mayankansys, no need of these error handling logic. Just fall back to the server implmenentation.

File: src/ansys/fluent/core/rest/client.py [hunk 25]
```diff
@@ -0,0 +1,763 @@
...
+        """Enter the context manager."""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        """Exit the context manager — calls :meth:`exit`."""
+        self.exit()
+
+
+
+"""
```
Comment (prmukherj) [inline src/ansys/fluent/core/rest/client.py]: @mayankansys, please remove this commented parts. If some of it is required please put it in a place suitable, probably top of this file or in the __init__.py

File: src/ansys/fluent/core/rest/rest_launcher.py [hunk 26]
```diff
@@ -0,0 +1,303 @@
...
+        )
+    )
+
+
+# ---------------------------------------------------------------------------
+# Public API — launchers
+# ---------------------------------------------------------------------------
+
+
+def launch_webserver(
```
Comment (prmukherj) [inline src/ansys/fluent/core/rest/rest_launcher.py]: @mayankansys, there should be a way to bypass most of this code in this method and just call pyfluent.launch_fluent with required additional arguments.
